### ISSUE TITLE
My thoughts for a new API

### ISSUE THREAD

**Issue by Nyholm** [mailgun/mailgun-php#186]
Here are my thoughts for the completing the [Better API milestone](https://github.com/mailgun/mailgun-php/milestone/1). 

I believe we can make the api client much easier to work with and much easier to maintain. I’ve been looking at 20+ big API clients and there are basically 3 ways you can make a client. 

1. Pure HTTP: Does only provide authentication and a `client::api(string $url, array $params)`. Like [Happyr/LinkedIn](https://github.com/Happyr/LinkedIn-API-client)

2. Defined Resources: Defined all API resources and separate different parts of the API into API\Foo classes. Each resource of the API has a method in the API\Foo class. Parameters to these methods would be scalars and arrays. Like [KNP/Github](https://github.com/KnpLabs/php-github-api).

3. Domain models: Like *Defined Resources* but we also use domain objects for API responses and for POSTing data. 

Usability and the work to maintain the API increases from 1 to 3. 

The clear drawback of using *Domain models* is that we have to update the client every time the API updates. Whenever the API adds a parameter in the request or response we need to do an update that reflects that API change. That is not the case with *Defined resources*

I suggest we take a lot of inspiration from KNP/Github and do something similar ourselves. 

-----------------------

It is important that we can make this update in a backwards compatible way. I believe it is possible. We should make sure to deprecate functions and add more features one at the time. 

-----------------------

I suggest the following new classes: 
* [ ] API\Message
* [ ] API\Domain
* [ ] API\Event
* [x] API\Stats
* [ ] API\Tag
* [ ] API\Suppression
* [ ] API\Route
* [ ] API\Webhook (Possibly API\Domain\Webhook)
* [ ] API\List
* [ ] API\Address
* [ ] API\Campaign

These classes reflect the [API reference](https://documentation.mailgun.com/api_reference.html).

I also suggest we create a `Maingun::getUrl` to be able to fetch **any** url. We do only populate the Request with authentication headers. This will solve issues with fetching paginated data (#128) and stored messages (#168). 

------------------

### Return types
Today we return a \stdClass. I suggest we return an array response with the data we get from the server. 

We should throw exceptions on non-200 HTTP status. We should also throw an `MissingArgumentEception` or `InvalidArugmentException` if we notice that the arguments are invalid before we send the request. 

------------------

I would like some feedback on this change before we get started. Please give a :+1: or :-1: and maybe share a comment or two.

**Comment by DavidGarciaCat** [comment]
Hey @Nyholm 

Here you are my impressions. Please feel free to discuss them:

### Classes

I think the new classes, based on the API Reference, is a good idea. It will help to identify easily how the API and this client is built.

Also, I would like to think an easy way to manage multiple domains using a single Mailgun instanced object, so maybe we should consider about an `API\Client` class, that should contain the API Key or the domain that we are going to use, so once we create a new Mailgun instance, we can set the expected `Client` depending on the actions that we want to run.

For example, my account is managing 6 domains, and may be an interesting option to re-use the same instantiated Mailgun object.

### Dependencies

Maybe we should consider to add inside the `composer.json` file (`require` key) all packages that we may need to use this SDK. Current `README.md` file contains this:
```
php composer.phar require mailgun/mailgun-php php-http/curl-client guzzlehttp/psr7
```

From my POV, when I need to add more than 1 vendor to install a package (`composer require vendor1 vendor2 vandor3`) I am trembling because I don't know if the `vendor2` or `vendor3` package will install a well tested version with the `vendor1`, so:
```
php composer.phar require mailgun/mailgun-php
```
should install automatically:
- `mailgun/mailgun-php`
- `php-http/curl-client`
- `guzzlehttp/psr7`

### Interfaces

Maybe is because I think on a big project instead of an SDK, but if we want to provide a flexible environment (and allow the user to extend it easily, if someone wants to add custom methods) maybe an interesting idea is to add Interfaces for each class that should contain a set of methods. Or - if we should not allow to extend the SDK - declare classes as `final`.

### Return types

I am sorry, but I disagree. PHP5 introduced OOP properly, and PHP7 has improved the OOP allowing more hint types (for example). Also, and based on my experience, work with objects is better than an array and, in the end, we always can do this:

```
$messages = $mailgun->Messages->getMessages();

foreach ($messages as $message) {
    // do some stuff here
}
```

Where:
- `$mailgun->Messages` is an instance of `API\Messages`
- `$messages` is an array of `API\Response\Message` objects
- `$message` is an instance of `API\Response\Message`

Why would you like to return an `array` instead of a `stdClass` or another object?

**Comment by Nyholm** [comment]
Thank you very much for your thoughts.

### Classes

I think we should reuse the same `Mailgun` and `HttpClient`. As long as you use 6 different **domains** and not API keys it would be no problem. 

### Dependencies

I disagree with that. What if a user does not like `guzzlehttp/psr7`. That implementation has some bugs and are not 100% compatible with PSR-7. (Partly my responsibility though... Im a maintainer to Guzzle...). We (Mailgun) should not prevent users for using another implementation. 

I know this is kind of new but it will increase flexibility. There are many other projects moving in this direction. PHPGeocoder, FosHttpCache, Stormpath, Facebook, Payum etc. 

### Interfaces

Agreed. We should let people be able to integrate easily. The problem with adding interfaces to the API classes is that we have to bump a major version if we add a method. 
It would be easy to add your own implementation of API\Foo because that is a standalone class. See example: https://github.com/KnpLabs/php-github-api/blob/2.0.0-rc4/lib/Github/Api/Issue.php

### Return types

I agree, classes are generally better than `\stdClass` and arrays. The issue is maintainability. Im not 100% sure here what we should do..

```php
// https://documentation.mailgun.com/api-stats.html
$result = $mailgun->api('stats')->total('my-domain', [
    'event' => array('accepted', 'delivered', 'failed'),
    'duration' => '1m'
]);
```

We could let `$result` be an array of the response ([see example](https://documentation.mailgun.com/api-stats.html#examples)) and let the end user do what ever. Or we could build an object representation of that array. But what would the win be?

Ping @sagikazarmark maybe you have some thoughts about the return types.

**Comment by DavidGarciaCat** [comment]
### Classes

👍 

### Dependencies

👍

I understand your point with the dependencies. I usually use - for example - Buzz HTTP Client because I think is a light and simple HTTP Client but I have used Guzzle too, and both of them are working well.

Maybe keeping this "open" we can provide a more flexible environment, specially if a project is already using a specific HTTP Client.

### Interfaces

👍

Then maybe we should think about Abstract classes - where needed - but not everywhere, of course.

### Return types

👍

I like the idea to build an object representation of that array.

**Comment by Nyholm** [comment]
> I like the idea to build an object representation of that array.

Are you talking about a `\stdClass` or should we define a class for each return type?

**Comment by DavidGarciaCat** [comment]
Well, an easy way to build a `\stdClass` is just run the `json_decode()` but that is a generic object, and does not contain a known map or references to each property. If we need to provide an object this is already offering it.

However, I think a DTO object is a better approach here. In the end, when we call the Message API endpoint (for example) we know the data that we are going to return (because we can already see it from the API Docs, right?), so probably define a DTO Class for each return type means more work now but more comprehensible and consistent results too.

**Comment by Nyholm** [comment]
You are probably correct. The only argument I have against that is high maintainability cost. Let's try it :)

**Comment by sagikazarmark** [comment]
Here are a few thoughts of mine:

## Classes & Interfaces

Personally I prefer having a proper domain model so that any transport detail is completely hidden from me. I also like the idea of having interfaces (actually I do it in of my clients). Here is a simple representation:

``` php
<?php

interface Client
{
    public function someResource() : SomeResource
}

interface SomeResource
{
    public function getResource(string $id) : Model;
}

interface Exception {}

final class Model
{
    private $id;

    public function getId() : string
    {
        if ($this->isNew()) {
            throw new class ('A new object does not have an ID') extends \LoginException implements Exception;
        }

        return $this->id;
    }

    public function isNew() : bool
    {
        return is_null($this->id);
    }
}

namespace Implementation
{
    final class Client implements \Client
    {
        public function someResource() : SomeResource
        {
            return new SomeResource($this->httpClient, $this->requestFactory, $this->serializer);
        }
    }

    final class SomeResource implements \SomeResource
    {
        public function getResource(string $id) : Model
        {
            // get the model and deserialize it
            return new \Model();
        }
    }
}
```

Explanation: There is a Client facade interface which holds the list of resources (eg. API\Message in @Nyholm 's example). The resource is the actual interface interacting with the API. These objects can also have dummy implementations eg. for unit tests. The models themselves are final classes as they are only DTOs, makes no sense to have alternative implementations. There is also an exception interface which must be implemented by all domain exceptions.

The only problem with the above interfaces is that you MUST break BC to add new API calls for example.

The above example is probably most "correct" from a typesafety, domain and logical point of view, it is also hard to maintain and more prone to break BC whenever the API changes. Since this is an official client you have somewhat control over the API's stability itself, or at least you are aware of the API's BC promise. So if you know the API is stable, no breaks without increasing a major version then go ahead with the above or a similar approach.


## Dependencies

I think the PSR-7 message implementation interoperability is not the main reason, rather HTTP Client one. Library users might already have a client installed, or they might have a PHP version which does not support the client that comes with this library, etc. There are a number of reasons why HTTP client interoperability is a good idea (the above two are the most common ones) and I think this kind of flexibility improves not just this library, but all the libraries having some kind of HTTP related task.

Also, this is not really different from optional dependencies. If you want to use a feature that relies on an optional package, you still have to install it. Although it is not optional in this case, this kind of procedure is known and accepted in other cases, so I don't see why it would be a problem here. And by using the virtual package, it is also quite straight forward and safe (application does not fail in production because of unmet dependencies).


## Return types

This is also partly covered in the classes section. I prefer having models, but it puts more maintenance and risk of breaking BC on the library. It requires some experimenting IMO.

**Comment by Nyholm** [comment]
Thank you Mark for your input. I agree and like what you are saying except:

> The only problem with the above interfaces is that you MUST break BC to add new API calls for example.

I believe we could be okey without the interfaces. Especially since the Client is very decoupled from the Mailgun Class. 

While reading your comment I came up with an idea. 

We should be able to provide different serializers. One Psr7ResponseSerializer, one ArraySerializer and one DomainModelSerializer. We could use them like: 


``` php
<?php
namespace Implementation
{

    final class SomeResource implements \SomeResource
    {
        public function getResource(string $id) : Model
        {
            $response = ...
            
            return $this->serializer->unseriize($response, \Model::class);
        }
    }
}
```

This will allow the users to chose the response they want from the API. The default should of course be the domain serializer.

**Comment by sagikazarmark** [comment]
@Nyholm here comes some of my thoughts about the recently merged #192:

## Naming

I am usually bad at naming, but I have a few ideas on this front:

- AbstractApi: instead of Abstract I would rather reflect somehow that it is the HTTP foundation of the API client. HttpApi? Abstract prefix is evil anyway
- ResponseDeserializer: it's quite weird that it is stored in a property called "serializer", also the implementations are called Serializers. Both could work for me, probably serializer sounds better, but be consistent.
- getSomeAPi methods: compared to the method names in API classes it sounds a little bit weird: you used there "total" and "all", logically eg. "stats" would be better, and IMO sounds better. When using: `$mailgun->stats()->total()` sounds cool.
- Response classes: do we need the Response suffix there?
- Item: without knowing the API itself: is Item a generic DTO or part of the domain model?
- CreatableFromArray: ArrayConstructable? ArraySerializable?
- ObjectSerializer: ModelSerializer? DomainSerializer?

## Response serialization

Although it is called serialization, as far as I can see it's normalization really. See this picture:

http://symfony.com/doc/current/_images/serializer_workflow.png

Not sure how this should be handled and if serialization and normalization should be separated within the library.

## Serializer concept

I really like this solution, it's actually very clever and flexible. However, my fear is that too much things are put into configuration and it might become hard to rely on the actual API. So my idea would be to somehow make this part of the API of the code: maybe have multiple resource implementations, each of them return the specific type of answer (eg. domain model or array). This could rely on the current code foundation, could even be generated. The point is that when using it in actual code it does not depend on the documentation what kind of answer we receive, because it's a little bit leaky. (I admit, it doesn't sound as good as I felt in my head 😄 )

**Comment by Nyholm** [comment]
Thank you for the input. I've created many smaller/more focused issues for your thoughts. I do agree with most of them. 

Im going to close this issue now and let the 7 issues be open.
